src/playermat/Playmat.ttslua

```diff
@@ -93,293 +70,271 @@ function onload(save_state)
+  local searchPosition = {-position[1], position[2], position[3] + 0.32}
+  local handler = makeDiscardHandlerFor(searchPosition, discardPosition)
+  local handlerName = 'handler' .. number
+  self.setVar(handlerName, handler)
+  self.createButton({
+    label = "Discard",
+    click_function = handlerName,
+    function_owner = self,
+    position = position,
+    scale = {0.12, 0.12, 0.12},
+    width = 800,
+    height = 280,
+    font_size = 180
+  })
 end
 
-function doNotReady(card)
-    if card.getVar("do_not_ready") == true then
-        return true
-    else
-        return false
-    end
+function findObjectsAtPosition(localPos)
+  return Physics.cast({
+    origin       = self.positionToWorld(localPos),
+    direction    = {0, 1, 0},
+    orientation  = {0, PLAY_ZONE_ROTATION.y + 90, 0},
+    type         = 3,
+    size         = {3.2, 1, 2},
+    max_distance = 0,
+    debug        = DEBUG
+  })
 end
 
--- draw X cards (shuffle discards if necessary)
-function drawCardsWithReshuffle(numCards)
-    if type(numCards) ~= "number" then numCards = 1 end
-
-    getDrawDiscardDecks()
-
-    if investigator == "Norman Withers" then
-        local harbinger = false
-        if topCard ~= nil and topCard.getName() == "The Harbinger" then
-            harbinger = true
-        else
-            if drawDeck ~= nil and not drawDeck.is_face_down then
-                local cards = drawDeck.getObjects()
-                local bottomCard = cards[#cards]
-                if bottomCard.name == "The Harbinger" then
-                    harbinger = true
-                end
-            end
-        end
+---------------------------------------------------------
+-- Upkeep button
+---------------------------------------------------------
 
-        if harbinger then
-            printToColor("The Harbinger is on top of your deck, not drawing cards", messageColor)
-            return -1
-        end
+function doUpkeep(_, color, alt_click)
+  setMessageColor(color)
 
-        if topCard ~= nil then
-            topCard.deal(numCards, PLAYER_COLOR)
-            numCards = numCards - 1
-            if numCards == 0 then return end
+  -- right-click binds to new player color
+  if alt_click then
+    PLAYER_COLOR = color
+    printToColor("Upkeep button bound to " .. color, color)
+    return
+  end
+  
+  local forcedLearning = false
+
+  -- unexhaust cards in play zone, flip action tokens and find forcedLearning
+  for _, v in ipairs(searchArea(PLAY_ZONE_POSITION, PLAY_ZONE_SCALE)) do
+    local obj = v.hit_object
+    if obj.tag == "Card" and not obj.is_face_down and not doNotReady(obj) then
+      local notes = JSON.decode(obj.getGMNotes()) or {}
+      if notes.id == "08031" then
+        forcedLearning = true
+      elseif notes.type ~= "Investigator" then
+        obj.setRotation(PLAY_ZONE_ROTATION)
+
+        -- check for cards with 'replenish' in their metadata
+        if notes.uses ~= nil then
+          local count = notes.uses[1].count
```

Comment:
Can we make this more robust, specifically by handling the multiple uses?  None of them have replenish right now but it's possible, would be nice to future-proof it

Reply:
I don't see the issue - isn't it "notes.uses" indepedant of the type of uses? Could you provide an example card (hypothetical) that wouldn't work?

Reply:
Uses is an array, this is only checking the first entry.  Currently Arcane Initiate (3) is the only card that has multiple use types, but if a second (or more) type had a replenish this would miss it

Reply:
Ah, that makes sense. So we probably wanna loop through all the entries in "uses" and check if there is replenish in one of them?

Reply:
Probably should just loop through and handle each uses independently.  Agree that it's unlikely, up to you if you want to fix it or not.  But since a lot of places are likely to (ahem) use uses, IMHO it's better to have them comprehensive so that when a new combination comes along we don't have to audit which (ahem) uses of uses handle which combinations.

Reply:
Would prefer leaving it like this because this case is not possible atm (and at the very earliest relevant with the nex investigator expansion which will be at least 6 months away)